### ISSUE TITLE
Licensing

### ISSUE THREAD

**Issue by devongovett** [solidjs-community/solid-aria#12]
Hey, I work on React Aria. Very happy to see this project extending our work to another framework! It's something I've wanted to experiment with for a long time, but haven't had time.

React Aria uses the Apache 2.0 license, which is very permissive but requires that the copyright headers in source files, [LICENSE](https://github.com/adobe/react-spectrum/blob/main/LICENSE), and [NOTICE.txt](https://github.com/adobe/react-spectrum/blob/main/NOTICE.txt) files be included in any redistributions or modifications. Since many of the files in this project appear to be direct ports of code from React Aria, we would really appreciate if you could include the license here as well.

Also, please let us know if there is anything we can do to help. Like I said, extending React Aria to support other frameworks is something we've had on our long term wishlist for a while. As you build out the primitives here, it would be really great if you could share any learnings with us, or ideas about how some of the code could be shared. Would be great to potentially come up with an abstraction where this stuff could be more easily shared between frameworks rather than duplicating efforts, porting bug fixes back and forth, etc.

**Comment by fabien-ml** [comment]
Hi, i've added the copyright header and NOTICE.txt.
For the license, Solid Aria should be Apache 2.0 too or it can stay MIT and put the React Aria Apache 2.0 license somewhere ?

Currently i'm just doing a copycat of React Aria (even the test files are almost the same) 😅. The major problem that i've found is related to `@react-stately/collections` and all collection related stuff, in Solid we can't rely on the JSX tree like its done in `CollectionBuilder`. Things like `React.cloneElement` can't be done too since there is no VDOM.

Also since it's just a port and i'm not the brain that "created" this original code, fixing bugs will be very challenging for me.
An abstraction for non frameworks related stuff would be cool like a `vanilla-aria` thing.

The Chakra UI team is also experimenting something similar with [zag.js](https://zagjs.com/) which uses finite state machines.

**Comment by fabien-ml** [comment]
@devongovett adding the copyright header to source files looks weird, it's look like all Solid Aria files are Adobe proprietary stuff.

Can we put something like this instead ? (like you do in [usePress](https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/interactions/src/usePress.ts#L13) for example when mentioning React code).

```js
// Portions of the code in this file are based on code from react-spectrum.
// Original licensing for the following can be found in the
// NOTICE file in the root directory of this source tree.
// See https://github.com/adobe/react-spectrum
```

**Comment by fabien-ml** [comment]
Also for a "framework agnostic" thing, a main problem is how we deal with internal state, for example React Stately heavily rely on react hooks (like [useToggleState](https://github.com/adobe/react-spectrum/blob/main/packages/%40react-stately/toggle/src/useToggleState.ts) does) and the SolidJS port does the same with primitives.

Things like React Aria [mergeProps](https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/utils/src/mergeProps.ts) is also different from the [Solid one](https://github.com/solidjs-community/solid-primitives/blob/main/packages/props/src/combineProps.ts)

**Comment by devongovett** [comment]
> The major problem that i've found is related to @react-stately/collections and all collection related stuff

Yeah, the collections API is certainly the most "React-y" part. But most parts of the code only rely on the [Collection interface](https://react-spectrum.adobe.com/react-stately/Collection.html). Our implementation builds a Collection from JSX, but you could implement an alternate version that uses plain JS objects or something. An important part we found was not trying to derive it from the rendered DOM, because when you get to things like virtualized scrolling, not all elements may be in the DOM. For keyboard navigation and such, we need the whole collection up front even if it isn't fully rendered. Some docs [here](https://react-spectrum.adobe.com/react-stately/collections.html) about the considerations there.

> Also for a "framework agnostic" thing, a main problem is how we deal with internal state, for example React Stately heavily rely on react hooks

Yep, that and the React event system are the main differences I found. For reference, I also experimented with [Vue](https://github.com/devongovett/vue-hooks) and [Svelte](https://github.com/devongovett/svelte-hooks) wrappers at one point.

One question as you're porting: do you think a good part of it could be automated somehow? Someone linked me to [mitosis](https://github.com/BuilderIO/mitosis), which is a compiler that can output components for a bunch of different frameworks. Wonder if a compiler approach could help convert hooks like this between frameworks as well? Obviously not 100% as the API would change in some places (like collections), but perhaps could help a good amount? Might be cool to experiment.

**Comment by fabien-ml** [comment]
Fixed by #14
